src/jmclient/wallet.py

```diff
@@ -1038,8 +1138,8 @@ def is_known_addr(self, addr):
         returns:
             bool
         """
-        script = self.addr_to_script(addr)
-        return script in self._script_map
+        assert isinstance(addr, str)
+        return addr in self._addr_map
```

Comment:
> Is this "type-check"assert necessary

Well, no, it's not. That's the reason why it's an assert. When running the python interpreter in optimization mode (`python -O`) it will strip out all asserts but during development you still get the benefit of additional sanity checks.

Reply:
Yeah, but by that logic we should then add `assert` in almost every function and method. Also, if we really want to do this, any reason to prefer `assert` over `TypeError`? I doubt the performance is at all noticeable.

Reply:
True, with the introduction of type hints this kind of assert can be expressed better in that way.

Reply:
Type hints are ignored at runtime. In order for the method to return a correct result, the address given must be of the expected type, or else the method will quietly return `False` even when the address *is* present when encoded in the expected type. Add a type hint, sure, but don't remove the runtime check.

Generally I don't support hard checks for preconditions that should always be true if the code is correct. Exceptions are for exceptional conditions. Incorrect code is not an exceptional condition; it's a bug. That's exactly what assertions are meant to catch.

Reply:
> Type hints are ignored at runtime

Yeah but this is Python. I just don't think adding `assert`/`Exception` everywhere is a scalable approach. I'm sure there are plenty of other cases where a wrong type would be a problem like here.

> Exceptions are for exceptional conditions. Incorrect code is not an exceptional condition; it's a bug.

That's reasonable, tho in this case, we don't know what the caller is gonna provide. It doesn't seem impossible to me that a caller somewhere might use a `try/except`.

Reply:
Notwithstanding C++'s `std::logic_error`, exception handling really isn't supposed to be used for *foreseeable* errors. A foreseeable error, by definition, is not exceptional. A caller shouldn't be wrapping a call in `try/except` to guard against its own misapplication of a called function.

Reply:
I'm not sure I follow. Aren't all exceptions foreseeable, or we wouldn't write an exception for it? In any case, terminology aside, there are plenty of (built-in) functions in Python that raise `TypeError`, and using `try/except` is very common.

Reply:
> Aren't all exceptions foreseeable, or we wouldn't write an exception for it?

Fair point. I should have said "avoidable." In the case of an error that will *only* arise if the *code* is incorrect, that is a case for an assertion. In the case of an error that may arise due to conditions outside of the programmer's control, that is a case for an exception. Pedantic ideals aside, I will grant you that there are many languages that (ab)use exceptions to signal programming errors, and I'm at all not surprised that Python is one of them.

I wouldn't raise a blocking objection if someone wanted to upgrade type-checking assertions to exceptions.